### PR TITLE
MySQL: Quote identifiers that include special characters

### LINKED ISSUE #59854: MySQL: cannot query datasets or tables names with hyphen
**What happened**:

If a data set as an hyphen, it is correctly displayed/selected in the dropdown selector, but when rendering the code the hyphen is considered as a minus operator. 

Screenshot 2022-12-05 at 20-43-31 Explore - MySQL _ home-automation - Grafana

**What you expected to happen**:
dataset and table name should be quoted for correct processing

**Environment**:
- Grafana version: 9.2.6
- Data source type & version: mariadb 10.6.11
- OS Grafana is installed on: FreeBSD
- User OS & Browser: FreeBSD / Firefox
- Grafana plugins:

- Others:

### PR DESCRIPTION
**Pull request by zoltanbedi**
This PR fixes an issue with MySQL where table names and identifiers in general were not handled when it included special characters or if it was a reserved word. Because of this we produced an invalid query. This PR fixes that issue by being smart and escaping/quoting those. See the reference for the rules https://dev.mysql.com/doc/refman/8.0/en/identifiers.html

**Which issue(s) does this PR fix?**:

Fixes #59854

### TODO

- [x] Fix code editor autocomplete
- [x] Add e2e tests
- [x] Document the change

### CODE REVIEW

File: public/app/features/plugins/sql/components/visual-query-builder/SelectRow.tsx [hunk 1]
```diff
@@ -137,6 +137,7 @@ export function SelectRow({ sql, format, columns, onSqlChange, functions }: Sele
             <EditorField label="Alias" optional width={15}>
               <Select
                 value={item.alias ? toOption(item.alias) : null}
+                inputId={`select-alias-${index}-${uniqueId()}`}
```
Comment (zoltanbedi) [inline public/app/features/plugins/sql/components/visual-query-builder/SelectRow.tsx:140]: This fixes an A11y check for the query editor.

File: public/app/features/plugins/sql/components/QueryHeader.tsx [hunk 2]
```diff
@@ -94,28 +92,14 @@ export function QueryHeader({
   return (
     <>
       <EditorHeader>
-        {/* Backward compatibility check. Inline select uses SelectContainer that was added in 8.3 */}
-        <ErrorBoundary
```
Comment (zoltanbedi) [inline public/app/features/plugins/sql/components/QueryHeader.tsx:98]: This was copied over from BigQuery and is not relevant here.

File: public/app/plugins/datasource/mysql/mySqlMetaQuery.ts [hunk 3]
```diff
@@ -1,60 +1,40 @@
...
 export function showDatabases() {
   return `SELECT DISTINCT TABLE_SCHEMA from information_schema.TABLES where TABLE_TYPE != 'SYSTEM VIEW' ORDER BY TABLE_SCHEMA`;
 }
 
-export function buildColumnQuery(queryModel: MySQLQueryModel, table: string, type?: string, timeColumn?: string) {
+export function buildColumnQuery(table: string, dbName?: string) {
   let query = 'SELECT column_name, data_type FROM information_schema.columns WHERE ';
-  query += buildTableConstraint(queryModel, table);
-
-  switch (type) {
```
Comment (zoltanbedi) [inline public/app/plugins/datasource/mysql/mySqlMetaQuery.ts:16]: This is not relevant anymore

File: public/app/plugins/datasource/mysql/sqlUtil.test.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { isValidIdentifier } from './sqlUtil';
+
+describe('isValidIdentifier', () => {
+  test.each([
+    { value: 'and', expected: false }, // Reserved keyword
+    { value: '1name', expected: false }, // Starts with value
+    { value: 'my-sql', expected: false }, // Contains not permitted character
+    { value: '$id', expected: false }, // $ sign shouldn't be the first character
+    { value: 'my sql', expected: false }, // Whitespace is not permitted
+    { value: 'mysql ', expected: false }, // Whitespace is not permitted at the end
```
Comment (baldm0mma) [inline public/app/plugins/datasource/mysql/sqlUtil.test.ts:10]: Any need to test whitespace prepending the string here?

Comment (zoltanbedi) [reply public/app/plugins/datasource/mysql/sqlUtil.test.ts:10]: yeah that can be added.

File: public/app/plugins/datasource/mysql/sqlCompletionProvider.ts [hunk 5]
```diff
@@ -1,22 +1,143 @@
...
+}
+
+export enum CustomSuggestionKind {
+  TablesWithinDatabase = 'tablesWithinDatabase',
+}
+
+export const customStatementPlacement: StatementPlacementProvider = () => [
+  {
+    id: CustomStatementPlacement.AfterDatabase,
+    //TODO: Figure out to only match this once
```
Comment (mdvictor) [inline public/app/plugins/datasource/mysql/sqlCompletionProvider.ts:39]: Is this still necessary?

File: public/app/plugins/datasource/mysql/sqlCompletionProvider.ts [hunk 6]
```diff
@@ -1,22 +1,143 @@
...
-      resolve: getMeta.current,
-    },
-    columns: {
-      resolve: getMeta.current,
-    },
+    customStatementPlacement,
+    customSuggestionKinds: customSuggestionKinds(getMeta),
   });
+
+export enum CustomStatementPlacement {
```
Comment (mdvictor) [inline public/app/plugins/datasource/mysql/sqlCompletionProvider.ts:28]: More of a nit: Is there any reason we don't just use constants here?

File: public/app/plugins/datasource/mysql/MySqlQueryModel.ts [hunk 7]
```diff
@@ -15,19 +15,21 @@ export default class MySQLQueryModel {
...
 
-  quoteIdentifier(value: string) {
+  static quoteIdentifier(value: string) {
     return '"' + value.replace(/"/g, '""') + '"';
   }
 
-  quoteLiteral(value: string) {
+  static quoteLiteral(value: string) {
     return "'" + value.replace(/'/g, "''") + "'";
   }
```
Comment (mdvictor) [inline public/app/plugins/datasource/mysql/MySqlQueryModel.ts:35]: Seems like the class can be refactored a bit: non-static method `getDatabase` and constructor aren't used anymore and class members also look like they can be deleted.

**Review by mdvictor — CHANGES_REQUESTED** [review]
This is awesome work, @zoltanbedi! Functionality works great! Tested queries with both hyphened table names and reserved words table names and they are looking good!

I left some comments with mostly nits. The biggest thing, imo, is the `MySqlQueryModel` refactor. Otherwise this looks great!

**Review by baldm0mma — APPROVED** [review]
fantastic!

**Review by mckn — COMMENTED** [review]
LGTM! Great work with this.

File: public/app/plugins/datasource/mysql/sqlCompletionProvider.ts (hunk 5 above)
Comment (zoltanbedi) [reply public/app/plugins/datasource/mysql/sqlCompletionProvider.ts]: It was thanks.

File: public/app/plugins/datasource/mysql/sqlCompletionProvider.ts (hunk 6 above)
Comment (zoltanbedi) [reply public/app/plugins/datasource/mysql/sqlCompletionProvider.ts]: not really.

File: public/app/plugins/datasource/mysql/MySqlQueryModel.ts (hunk 7 above)
Comment (zoltanbedi) [reply public/app/plugins/datasource/mysql/MySqlQueryModel.ts]: I agree.

**Comment by zoltanbedi** [comment]
@mdvictor thanks for the review. Addressed your feedbacks.

**Review by mdvictor — APPROVED** [review]
LGTM! 

Now that I'm looking at the `MySqlQueryModel` after the refactor, maybe it's worth moving those 2 functions in `sqlUtil` and get rid of this file alltogether? What do you think?

**Comment by zoltanbedi** [comment]
> Now that I'm looking at the `MySqlQueryModel` after the refactor, maybe it's worth moving those 2 functions in `sqlUtil` and get rid of this file alltogether? What do you think?

Absolutely!

**Pull request merged**